core/state/access_list.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type accessList struct {
+	addresses map[common.Address]int
+	slots     []map[common.Hash]bool
+}
```

Comment:
IMHO this `map` + `slice` combo makes things more complicated than they should be. Wouldn't using a `map[common.Address]map[common.Hash]struct{}` be simpler? Yes, it looks a bit unwieldy, but code handling wise IMHO almost all the code below could be simplified.

Reply:
I'm not sure if the `bool` has any meaning other than "present". If only that, `struct{}` is better because it doesn't use any memory to store an extra `true` value.

Reply:
Yes, it would be easier. I was just trying to keep the memory down. I can change it if you prefer

Reply:
I'd still have made it even smaller :)) 

```
type accessList map[common.Address]map[common.Hash]struct{}
```

That would have avoided the need for the `-1` index special marker.